rfc/kubernetes-operator.md

```diff
@@ -0,0 +1,89 @@
+
+## Design
+
+See https://docs.google.com/document/d/1Fs-xK7sOZ4YLSw1ae_cQ_rsB0kMNon9AfYli2a5Udhs/edit#heading=h.7vexvfs9sl5e
+
+### Dependencies
+
+No new dependencies to existing projects. Some (relatively small) changes are necessary in Halyard to:
+
+1) Generate manifests from a dynamic config (not in `~/.hal/`) at runtime
+2) Generate CR and `configMap` from an existing config to facilitate the migration to an operator based deployment.
+
+
+## Drawbacks
+
+It's another project. The operator is also a new deployment running in the user's cluster. If the operator is strictly a tool to configure (vs monitor), that may seem overkill.
+
+## Prior Art and Alternatives
+
+We considered that [operator](https://operatorhub.io/operator/spinnaker-operator) but couldn't find good documentation, the code, or users using it. It also seems to be running a Halyard job in the cluster. Feel free to correct me.
+
+## Known Unknowns
+
+The `spec` of the CRD is still in flux. As we add more features, it will need to evolve before being stable. As an example, new fields are needed to track validation options. The existing `status` is also a first stab.
+
+We'd love to get feedback on the user workflows that this enables (or that are lacking). We already heard about support for Kubernetes secret as a common requirement.
+
+
+## Security, Privacy, and Compliance
+
+Security of the operator is controlled via a `ClusterRole`. The current definition may need to be tightened.
+
+## Operations
+
+After installing the Operator in a target cluster, the user can manage `SpinnakerService` using `kubectl` and manifests. How do you get these manifests?
+
+1) Users can maintain the `configMap` by hand. We’ll develop documentation to make it easier. We also expect (and have seen) users scripting the Halyard configuration with helm or kustomize.
+2) We added an endpoint to Halyard to generate the `configMap` and `SpinnakerService` from an existing Halyard configuration. This can be used as a one time migration to using configMaps or as a `hal deploy manifests | kubectl apply -f -`
+3) We can add an `Operator` deployment type to Halyard that applies a generated `configMap` and `SpinnakerService` - similar to how `Distributed` type manifests are applied. In this scenario, the user can keep a local Halyard but deploy with the Operator under the hood.
+
+There's currently no telemetry baked in.
+
+## Risks
+
+The implementation relies on Halyard's deployer being able to generate manifests instead of applying them.
+
+## Future Possibilities
+
+- More preflight and scheduled checks for connectivity, authorization: we'd for instance like to move some of the cloud provider account check to preflight check to speed up Clouddriver startup. The health of Spinnaker could be visible via `kubectl describe` and events.
```

Comment:
As the operator performs more and more preflight checks, it will need to understand more and more about the implementation details of each microservice.  How do you see this being implemented in a way that still maintains separation of concerns?

Will the operator need to know how to talk to a Jenkins server so that it can preflight things before spinning up Igor?  (As one example.)

Reply:
That's the idea. I realize it's no small feat and getting 100% preflight check will take a long time but we believe we can achieve address the most common issues reported relatively fast.